src/net_processing.cpp

```diff
@@ -4286,50 +4315,72 @@ class CompareInvMempoolOrder
-    const CNetMsgMaker msgMaker(pto->GetCommonVersion());
+    // Use mockable time for ping timeouts.
+    // This means that setmocktime may cause pings to time out.
+    auto now = GetTime<std::chrono::microseconds>();
+
+    if (gArgs.GetArg("-pingtimeout", DEFAULT_PING_TIMEOUT) &&
+        peer.m_ping_nonce_sent &&
+        now > peer.m_ping_start.load() + std::chrono::seconds{TIMEOUT_INTERVAL}) {
+        LogPrint(BCLog::NET, "ping timeout: %fs peer=%d\n", 0.000001 * count_microseconds(now - peer.m_ping_start.load()), peer.m_id);
+        node_to.fDisconnect = true;
+        return;
+    }
 
-    //
-    // Message: ping
-    //
+    const CNetMsgMaker msgMaker(node_to.GetCommonVersion());
     bool pingSend = false;
-    if (pto->fPingQueued) {
+
+    if (peer.m_ping_queued) {
         // RPC ping request by user
         pingSend = true;
     }
-    if (pto->nPingNonceSent == 0 && pto->m_ping_start.load() + PING_INTERVAL < GetTime<std::chrono::microseconds>()) {
+
+    if (peer.m_ping_nonce_sent == 0 && now > peer.m_ping_start.load() + PING_INTERVAL) {
         // Ping automatically sent as a latency probe & keepalive.
         pingSend = true;
     }
+
     if (pingSend) {
         uint64_t nonce = 0;
         while (nonce == 0) {
             GetRandBytes((unsigned char*)&nonce, sizeof(nonce));
         }
-        pto->fPingQueued = false;
-        pto->m_ping_start = GetTime<std::chrono::microseconds>();
-        if (pto->GetCommonVersion() > BIP0031_VERSION) {
-            pto->nPingNonceSent = nonce;
-            m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::PING, nonce));
+        peer.m_ping_queued = false;
+        peer.m_ping_start = GetTime<std::chrono::microseconds>();
+        if (node_to.GetCommonVersion() > BIP0031_VERSION) {
+            peer.m_ping_nonce_sent = nonce;
+            m_connman.PushMessage(&node_to, msgMaker.Make(NetMsgType::PING, nonce));
         } else {
             // Peer is too old to support ping command with nonce, pong will never arrive.
-            pto->nPingNonceSent = 0;
-            m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::PING));
+            peer.m_ping_nonce_sent = 0;
+            m_connman.PushMessage(&node_to, msgMaker.Make(NetMsgType::PING));
         }
     }
+}
+
+bool PeerManagerImpl::SendMessages(CNode* pto)
+{
+    // We must call MaybeDiscourageAndDisconnect first, to ensure that we'll
+    // disconnect misbehaving peers even before the version handshake is complete.
+    if (MaybeDiscourageAndDisconnect(*pto)) return true;
+
+    // Don't send anything until the version handshake is complete
+    if (!pto->fSuccessfullyConnected || pto->fDisconnect)
+        return true;
+
+    PeerRef peer = GetPeerRef(pto->GetId());
+
+    MaybeSendPing(*pto, *peer);
+
+    // MaybeSendPing may have marked peer for disconnection
+    if (pto->fDisconnect) return true;
+
+    // If we get here, the outgoing message serialization version is set and can't change.
```

Comment:
this comment previously followed the `fSuccesfullyConnected` check, which I think(?) was the intent -> `fSuccesfullyConnected` is set when processing the `VERACK` message, and `m_greatest_common_version` gets set when processing the `VERSION` message.

but now this comment is below `MaybeSendPing`, which also might retrieve `m_greatest_common_version` to construct the ping message. I find the intent of this comment to be less clear after these changes.

Reply:
This is a comment on the line below, which is contructing a `CNetMsgMaker` object with the correct serialization version:

`const CNetMsgMaker msgMaker(pto->GetCommonVersion());`

Reply:
I've moved this back to immediately after the `fSuccessfullyConnected` check. Hopefully that's clearer.